README.txt
Please see our on-line docs for detailed instructions, examples, support forums, etc.:

http://simplegeo.com/docs/clients-code-libraries/python


See also these other two libraries:


https://github.com/simplegeo/python-simplegeo-context

http://pypi.python.org/pypi/simplegeo-context


https://github.com/simplegeo/python-simplegeo-places

http://pypi.python.org/pypi/simplegeo-places


debian/changelog
python-simplegeo-shared (2.5.127) maverick; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 13 Dec 2011 17:39:08 +0000

python-simplegeo-shared (2.5.126) maverick; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 09 Dec 2011 01:11:39 +0000

python-simplegeo-shared (2.5.85) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 02 May 2011 23:49:03 +0000

python-simplegeo-shared (2.5.84) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 15 Apr 2011 21:21:39 +0000

python-simplegeo-shared (2.5.83) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 05 Apr 2011 23:12:11 +0000

python-simplegeo-shared (2.5.82) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 17 Feb 2011 21:54:02 +0000

python-simplegeo-shared (2.5.81) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * add test that shows Feature class rejects coords which aren't
    numbers when it is constructed So this is another thing that could
    have caused the misbehavior Wade saw -- a client sending GeoJSON
    with strings for coords -- but didn't.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 10 Feb 2011 22:59:00 +0000

python-simplegeo-shared (2.5.80) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * add test that the geojson produced has coords as numbers, not as
    strings Wade reported that our python client is submitting coords as
    strings. This test shows that the problem isn't the Feature class
    (which uses pyutil.jsonutil) serializing the coords incorrectly.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 10 Feb 2011 22:47:00 +0000

python-simplegeo-shared (2.5.79) lucid; urgency=low

  [ Ben Standefer ]
  * Add annotations endpoint.
  * Add annotations endpoint.
  * Add validation and tests for annotations.
  * Fix function names in annotations tests.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 08 Feb 2011 17:31:57 +0000

python-simplegeo-shared (2.5.78) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 04 Feb 2011 17:52:10 +0000

python-simplegeo-shared (2.5.77) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * update version to 2.5
  * make the error message on input validation easier to understand
    (thanks to Kim for reporting this) loosen to_unicode() to handle
    utf-8 encoded bytes instead of requiring ascii

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 04 Feb 2011 17:40:56 +0000

python-simplegeo-shared (2.4.72) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * change version number from 2.3 to 2.4

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 22:16:34 +0000

python-simplegeo-shared (2.3.71) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 22:10:31 +0000

python-simplegeo-shared (2.3.70) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * raise the requirement on the version of python-oauth2 to v1.5 That
    version has better handling unicode which we rely on.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 22:09:34 +0000

python-simplegeo-shared (2.3.69) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 22:03:44 +0000

python-simplegeo-shared (2.3.68) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 20:28:37 +0000

python-simplegeo-shared (2.3.67) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 19:50:34 +0000

python-simplegeo-shared (2.3.66) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 18:36:39 +0000

python-simplegeo-shared (2.3.65) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 17:42:00 +0000

python-simplegeo-shared (2.3.64) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 17:34:45 +0000

python-simplegeo-shared (2.3.63) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 17:22:21 +0000

python-simplegeo-shared (2.3.62) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:43:31 +0000

python-simplegeo-shared (2.3.60) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:32:22 +0000

python-simplegeo-shared (2.3.59) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 18 Jan 2011 17:36:06 +0000

python-simplegeo-shared (2.3.58) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * add "private" field to properties, default to False, add doc
  * raise instructive exception if the data argument is neither a
    unicode argument nor a str containing an ascii-encoded string
  *  * loosen validation on longitudes to allow longitudes up to 360
    degrees away from the Prime Meridian (optionally -- you can request
    stricter validation)  * do lat/lon validation on Feature.from_dict()
    * validation failures raise TypeError, not AssertionError, and
    validation proceeds even if PYTHONOPTIMIZE  * add unit test of
    to_unicode()  * raise version number to 2.1
  * fix test cases after merge

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 18 Jan 2011 17:33:15 +0000

python-simplegeo-shared (2.2.57) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 18 Jan 2011 16:43:58 +0000

python-simplegeo-shared (2.2.56) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 18 Jan 2011 16:41:00 +0000

python-simplegeo-shared (2.2.55) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * raise instructive exception if the data argument is neither a

simplegeo/shared/test/test_record.py
import unittest
import re
from simplegeo.shared import Feature, deep_swap
from decimal import Decimal as D

class FeatureTest(unittest.TestCase):

    def test_geojson_is_correct(self):
        f = Feature(coordinates=[-90, D('171.0')], properties={'record_id': 'my_id'}, strict_lon_validation=True)
        stringy = f.to_json()
        self.failUnlessEqual(stringy, '{"geometry": {"type": "Point", "coordinates": [171.0, -90]}, "type": "Feature", "id": null, "properties": {"record_id": "my_id", "private": false}}')

    def test_swapper(self):
        t1 = (2, 1)
        self.failUnlessEqual(deep_swap(t1), (1, 2))

        linestring1 = [(2, 1), (4, 3), (6, 5), (8, 7)]
        self.failUnlessEqual(
            deep_swap(linestring1),
            [(1, 2), (3, 4), (5, 6), (7, 8)]
            )

        multipolygon1 = [
            [[[102.0, 2.0], [103.0, 2.0], [103.0, 3.0], [102.0, 3.0], [102.0, 2.0]]],
            [[[100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0]],
             [[100.2, 0.2], [100.8, 0.2], [100.8, 0.8], [100.2, 0.8], [100.2, 0.2]]]
            ]

        self.failUnlessEqual(deep_swap(multipolygon1),
                             [[[(2.0, 102.0), (2.0, 103.0), (3.0, 103.0), (3.0, 102.0), (2.0, 102.0)]], [[(0.0, 100.0), (0.0, 101.0), (1.0, 101.0), (1.0, 100.0), (0.0, 100.0)], [(0.2, 100.2), (0.2, 100.8), (0.8, 100.8), (0.8, 100.2), (0.2, 100.2)]]]
                             )

    def test_record_constructor_useful_validation_error_message(self):
        try:
            Feature(coordinates=[181, D('10.0')], properties={'record_id': 'my_id'})
        except TypeError, e:
            self.failUnless(str(e).startswith('The first argument'), str(e))
            self.failUnless('is required to be a 2-element sequence' in str(e), str(e))
        else:
            self.fail('Should have raised exception.')

        try:
            Feature(coordinates=[-90, D('181.0')], properties={'record_id': 'my_id'}, strict_lon_validation=True)
        except TypeError, e:
            self.failUnless('181' in str(e), str(e))
        else:
            self.fail('Should have raised exception.')

        try:
            Feature(coordinates=[-90, D('361.0')], properties={'record_id': 'my_id'})
        except TypeError, e:
            self.failUnless('361' in str(e), str(e))
        else:
            self.fail('Should have raised exception.')

        try:
            Feature(coordinates=['-90', D('361.0')], properties={'record_id': 'my_id'})
        except TypeError, e:
            err_msg_re = re.compile("argument is required to be.*number.*not: .*<type 'str'")
            self.failUnless(err_msg_re.search(str(e)), str(e))
        else:
            self.fail('Should have raised exception.')

    def test_record_constructor(self):
        self.failUnlessRaises(TypeError, Feature, D('11.0'), D('10.0'), properties={'record_id': 'my_id'})

        # lat exceeds bound
        self.failUnlessRaises(TypeError, Feature, (D('91.0'), D('10.1')), properties={'record_id': 'my_id'})

        # lon exceeds bound
        self.failUnlessRaises(TypeError, Feature, (D('10.1'), D('180.1')), properties={'record_id': 'my_id'}, strict_lon_validation=True)

        record = Feature(coordinates=(D('11.0'), D('10.0')), properties={'record_id': 'my_id'})
        self.failUnlessEqual(record.properties.get('record_id'), 'my_id')
        self.failUnlessEqual(record.id, None)
        self.failUnlessEqual(record.geomtype, 'Point')
        self.failUnlessEqual(record.coordinates[0], D('11.0'))
        self.failUnlessEqual(record.coordinates[1], D('10.0'))

        record = Feature(coordinates=(D('11.0'), D('10.0')), simplegeohandle='SG_abcdefghijklmnopqrstuv')
        self.failUnlessEqual(record.properties.get('record_id'), None)
        self.failUnlessEqual(record.id, 'SG_abcdefghijklmnopqrstuv')

        record = Feature(coordinates=(D('11.0'), D('10.0')), properties={'record_id': 'my_id'}, simplegeohandle='SG_abcdefghijklmnopqrstuv')
        self.failUnlessEqual(record.properties.get('record_id'), 'my_id')
        self.failUnlessEqual(record.id, 'SG_abcdefghijklmnopqrstuv')

        record = Feature(coordinates=(D('11.0'), D('10.0')))
        self.failUnlessEqual(record.properties.get('record_id'), None)
        self.failUnlessEqual(record.id, None)

        record = Feature((D('11.0'), D('10.0')), properties={'record_id': 'my_id'})
        self.failUnlessEqual(record.properties.get('record_id'), 'my_id')

        record = Feature((11.0, 10.0), properties={'record_id': 'my_id'})
        self.failUnlessEqual(record.geomtype, 'Point')
        self.failUnlessEqual(record.coordinates[0], 11.0)
        self.failUnlessEqual(record.coordinates[1], 10.0)
        self.failUnlessEqual(record.properties.get('record_id'), 'my_id')

        record = Feature([[(11.0, 179.9), (12, -179.9)]], geomtype='Polygon')
        self.failUnlessEqual(record.geomtype, 'Polygon')
        self.failUnlessEqual(len(record.coordinates[0]), 2)
        self.failUnlessEqual(record.coordinates[0][0], (11.0, 179.9))

        jsondict = record.to_dict()
        self.failUnlessEqual(jsondict['geometry']['coordinates'][0][0], (179.9, 11.))

    def test_record_from_dict_lon_validation(self):
        record_dict = {
                     'geometry' : {
                                   'type' : 'Point',
                                   'coordinates' : [D('181.0'), D('11.0')]
                                   },
                     'type' : 'Feature',
                     'properties' : {
                                     'record_id' : 'my_id',
                                     'key' : 'value'  ,
                                     'type' : 'object'
                                     }
                     }

        self.failUnlessRaises(TypeError, Feature.from_dict, record_dict, True)
        record = Feature.from_dict(record_dict, strict_lon_validation=False)
        self.assertEquals(record.coordinates[0], D('11.0'))
        self.assertEquals(record.coordinates[1], D('181.0'))
        self.assertEquals(record.properties.get('record_id'), 'my_id')
        self.assertEquals(record.properties['key'], 'value')
        self.assertEquals(record.properties['type'], 'object')

    def test_record_from_dict(self):
        record_dict = {
                     'geometry' : {
                                   'type' : 'Point',
                                   'coordinates' : [D('10.0'), D('11.0')]
                                   },
                     'type' : 'Feature',
                     'properties' : {
                                     'record_id' : 'my_id',
                                     'key' : 'value'  ,
                                     'type' : 'object'
                                     }
                     }

        record = Feature.from_dict(record_dict)
        self.assertEquals(record.coordinates[0], D('11.0'))
        self.assertEquals(record.coordinates[1], D('10.0'))
        self.assertEquals(record.properties.get('record_id'), 'my_id')
        self.assertEquals(record.properties['key'], 'value')
        self.assertEquals(record.properties['type'], 'object')

        self.assertEquals(record.to_json(), '{"geometry": {"type": "Point", "coordinates": [10.0, 11.0]}, "type": "Feature", "id": null, "properties": {"record_id": "my_id", "type": "object", "private": false, "key": "value"}}')

        record_dict = {
                     'geometry' : {
                                   'type' : 'Point',
                                   'coordinates' : [D('10.0'), D('11.0')]
                                   },
                     'id' : 'SG_abcdefghijklmnopqrstuv',
                     'type' : 'Feature',
                     'properties' : {
                                     'key' : 'value'  ,
                                     'type' : 'object'
                                     }
                     }

        record = Feature.from_dict(record_dict)
        self.assertEquals(record.properties.get('record_id'), None)
        self.assertEquals(record.id, 'SG_abcdefghijklmnopqrstuv')

        record_dict = {
                     'geometry' : {
                                   'type' : 'Point',
                                   'coordinates' : [D('10.0'), D('11.0')]
                                   },
                     'id' : 'SG_abcdefghijklmnopqrstuv',
                     'type' : 'Feature',
                     'properties' : {
                                     'record_id' : 'my_id',
                                     'key' : 'value'  ,
                                     'type' : 'object'
                                     }
                     }

        record = Feature.from_dict(record_dict)
        self.assertEquals(record.properties.get('record_id'), 'my_id')
        self.assertEquals(record.id, 'SG_abcdefghijklmnopqrstuv')

        record_dict = {
                     'geometry' : {
                                   'type' : 'Point',
                                   'coordinates' : [10.0, 11.0]
                                   },
                     'type' : 'Feature',
                     'properties' : {
                                     'record_id' : 'my_id',
                                     'key' : 'value'  ,
                                     'type' : 'object'
                                     }
                     }

        record = Feature.from_dict(record_dict)
        self.assertEquals(record.coordinates[0], 11.0)
        self.assertEquals(record.coordinates[1], 10.0)

    def test_record_to_dict_sets_id_correctly(self):
        handle = 'SG_abcdefghijklmnopqrstuv'
        record_id = 'this is my record #1. my first record. and it is mine'
        rec = Feature(coordinates=(D('11.03'), D('10.03')), simplegeohandle=handle, properties={'record_id': record_id})
        dic = rec.to_dict()
        self.failUnlessEqual(dic.get('id'), handle)
        self.failUnlessEqual(dic.get('properties', {}).get('record_id'), record_id)

        rec = Feature(coordinates=(D('11.03'), D('10.03')), simplegeohandle=handle, properties={'record_id': None})
        dic = rec.to_dict()
        self.failUnlessEqual(dic.get('id'), handle)
        self.failUnlessEqual(dic.get('properties', {}).get('record_id'), None)

        rec = Feature(coordinates=(D('11.03'), D('10.03')), simplegeohandle=handle, properties={'record_id': None})
        dic = rec.to_dict()
        self.failUnlessEqual(dic.get('id'), handle)
        self.failUnlessEqual(dic.get('properties', {}).get('record_id'), None)

        rec = Feature(coordinates=(D('11.03'), D('10.03')), simplegeohandle=None, properties={'record_id': None})
        dic = rec.to_dict()
        self.failUnlessEqual(dic.get('id'), None)
        self.failUnlessEqual(dic.get('properties', {}).get('record_id'), None)


simplegeo/shared/__init__.py
from _version import __version__

API_VERSION = '1.0'

import copy, re
from decimal import Decimal as D

from httplib2 import Http
import oauth2 as oauth

import ipaddr

from urlparse import urljoin

from pyutil import jsonutil as json
from pyutil.assertutil import precondition, _assert

# example: http://api.simplegeo.com/1.0/feature/abcdefghijklmnopqrstuvwyz.json

def json_decode(jsonstr):
    try:
        return json.loads(jsonstr)
    except (ValueError, TypeError), le:
        raise DecodeError(jsonstr, le)

def swap(tupleab):
    return (tupleab[1], tupleab[0])

def deep_swap(struc):
    if is_numeric(struc[0]):
        _assert (len(struc) == 2, (type(struc), repr(struc)))
        _assert (is_numeric(struc[1]), (type(struc), repr(struc)))
        return swap(struc)
    return [deep_swap(sub) for sub in struc]

def deep_validate_lat_lon(struc, strict_lon_validation=False):
    """
    For the meaning of strict_lon_validation, please see the function
    is_valid_lon().
    """
    if not isinstance(struc, (list, tuple, set)):
        raise TypeError('argument is required to be a sequence (of sequences of...) numbers, not: %s :: %s' % (struc, type(struc)))
    if is_numeric(struc[0]):
        if not len(struc) == 2:
            raise TypeError("The leaf element of this structure is required to be a tuple of length 2 (to hold a lat and lon).")

        _assert_valid_lat(struc[0])
        _assert_valid_lon(struc[1], strict=strict_lon_validation)
    else:
        for sub in struc:
            deep_validate_lat_lon(sub, strict_lon_validation=strict_lon_validation)
    return True

SIMPLEGEOHANDLE_RSTR=r"""SG_[A-Za-z0-9]{22}(?:_-?[0-9]{1,3}(?:\.[0-9]+)?_-?[0-9]{1,3}(?:\.[0-9]+)?)?(?:@[0-9]+)?$"""
SIMPLEGEOHANDLE_R= re.compile(SIMPLEGEOHANDLE_RSTR)
def is_simplegeohandle(s):
    return isinstance(s, basestring) and SIMPLEGEOHANDLE_R.match(s)

FEATURES_URL_R=re.compile("http://(.*)/features/([A-Za-z_,-]*).json$")

def is_numeric(x):
    return isinstance(x, (int, long, float, D))

def is_valid_lat(x):
    return is_numeric(x) and (x <= 90) and (x >= -90)

def _assert_valid_lat(x):
    if not is_valid_lat(x):
        raise TypeError("not a valid lat: %s" % (x,))

def _assert_valid_lon(x, strict=False):
    if not is_valid_lon(x, strict=strict):
        raise TypeError("not a valid lon (strict=%s): %s" % (strict, x,))

def is_valid_lon(x, strict=False):
    """
    Longitude is technically defined as extending from -180 to
    180. However in practice people sometimes prefer to use longitudes
    which have "wrapped around" past 180. For example, if you are
    drawing a polygon around modern-day Russia almost all of it is in
    the Eastern Hemisphere, which means its longitudes are almost all
    positive numbers, but the easternmost part of it (Big Diomede
    Island) lies a few degrees east of the International Date Line,
    and it is sometimes more convenient to describe it as having
    longitude 190.9 instead of having longitude -169.1.

    If strict=True then is_valid_lon() requires a number to be in
    [-180..180] to be considered a valid longitude. If strict=False
    (the default) then it requires the number to be in [-360..360].
    """
    if strict:
        return is_numeric(x) and (x <= 180) and (x >= -180)
    else:
        return is_numeric(x) and (x <= 360) and (x >= -360)

def to_unicode(s):
    """ Convert to unicode, raise exception with instructive error
    message if s is not unicode, ascii, or utf-8. """
    if not isinstance(s, unicode):
        if not isinstance(s, str):
            raise TypeError('You are required to pass either unicode or string here, not: %r (%s)' % (type(s), s))
        try:
            s = s.decode('utf-8')
        except UnicodeDecodeError, le:
            raise TypeError('You are required to pass either a unicode object or a utf-8 string here. You passed a Python string object which contained non-utf-8: %r. The UnicodeDecodeError that resulted from attempting to interpret it as utf-8 was: %s' % (s, le,))
    return s

class Feature:
    def __init__(self, coordinates, geomtype='Point', simplegeohandle=None, properties=None, strict_lon_validation=False):
        """
        The simplegeohandle and the record_id are both optional -- you
        can have one or the other or both or neither.

        A simplegeohandle is globally unique and is assigned by the
        Places service. It is returned from the Places service in the
        response to a request to add a place to the Places database
        (the add_feature method).

        The simplegeohandle is passed in as an argument to the
        constructor, named "simplegeohandle", and is stored in the
        "id" attribute of the Feature instance.

        A record_id is scoped to your particular user account and is
        chosen by you. The only use for the record_id is in case you
        call add_feature and you have already previously added that
        feature to the database -- if there is already a feature from
        your user account with the same record_id then the Places
        service will return that feature to you, along with that
        feature's simplegeohandle, instead of making a second, duplicate
        feature.

        A record_id is passed in as a value in the properties dict
        named "record_id".

        geomtype is a GeoJSON geometry type such as "Point",
        "Polygon", or "Multipolygon". coordinates is a GeoJSON
        coordinates *except* that each lat/lon pair is written in
        order lat, lon instead of the GeoJSON order of lon, at.

        When a Feature is being submitted to the SimpleGeo Places
        database, if there is a key 'private' in the properties dict
        which is set to True, then the Feature is intended to be
        visible only to your user account. If there is no 'private'
        key or if there is a 'private' key which is set to False, then
        the Feature is intended to be merged into the publicly visible
        Places Database.

        Note that even if it is intended to be merged into the public
        Places Database the actual process of merging it into the
        public shared database may take some time, and the newly added
        Feature will be visible to your account right away even if it
        isn't (yet) visible to the public.

        For the meaning of strict_lon_validation, please see the
        function is_valid_lon().
        """
        try:
            deep_validate_lat_lon(coordinates, strict_lon_validation=strict_lon_validation)
        except TypeError, le:
            raise TypeError("The first argument, 'coordinates' is required to be a 2-element sequence of lon, lat for a point (or a more complicated set of coordinates for polygons or multipolygons), but it was %s :: %r. The error that was raised from validating this was: %s" % (type(coordinates), coordinates, le))

        if not (simplegeohandle is None or is_simplegeohandle(simplegeohandle)):
            raise TypeError("The third argument, 'simplegeohandle' is required to be None or to match this regex: %s, but it was %s :: %r" % (SIMPLEGEOHANDLE_RSTR, type(simplegeohandle), simplegeohandle))

        record_id = properties and properties.get('record_id') or None
        if not (record_id is None or isinstance(record_id, basestring)):
            raise TypeError("record_id is required to be None or a string, but it was: %r :: %s." % (type(record_id), record_id))
        self.strict_lon_validation = strict_lon_validation
        precondition(coordinates)
        self.id = simplegeohandle
        self.coordinates = coordinates
        self.geomtype = geomtype
        self.properties = {'private': False}
        if properties:
            self.properties.update(properties)

    @classmethod
    def from_dict(cls, data, strict_lon_validation=False):
        """
        data is a GeoJSON standard data structure, including that the
        coordinates are in GeoJSON order (lon, lat) instead of
        SimpleGeo order (lat, lon)
        """
        assert isinstance(data, dict), (type(data), repr(data))
        coordinates = deep_swap(data['geometry']['coordinates'])
        try:
            deep_validate_lat_lon(coordinates, strict_lon_validation=strict_lon_validation)
        except TypeError, le:
            raise TypeError("The 'coordinates' value is required to be a 2-element sequence of lon, lat for a point (or a more complicated set of coordinates for polygons or multipolygons), but it was %s :: %r. The error that was raised from validating this was: %s" % (type(coordinates), coordinates, le))
        feature = cls(
            simplegeohandle = data.get('id'),
            coordinates = coordinates,
            geomtype = data['geometry']['type'],
            properties = data.get('properties')
            )

        return feature

    def to_dict(self):
        """
        Returns a GeoJSON object, including having its coordinates in
        GeoJSON standad order (lon, lat) instead of SimpleGeo standard
        order (lat, lon).
        """
        return {
            'type': 'Feature',
            'id': self.id,
            'geometry': {
                'type': self.geomtype,
                'coordinates': deep_swap(self.coordinates)
            },
            'properties': copy.deepcopy(self.properties),
        }

    @classmethod
    def from_json(cls, jsonstr):
        return cls.from_dict(json_decode(jsonstr))

    def to_json(self):
        return json.dumps(self.to_dict())


class Client(object):
    realm = "http://api.simplegeo.com"
    endpoints = {
        'feature': 'features/%(simplegeohandle)s.json',
        'annotations': 'features/%(simplegeohandle)s/annotations.json',
    }

    def __init__(self, key, secret, api_version=API_VERSION, host="api.simplegeo.com", port=80):
        self.host = host
        self.port = port
        self.consumer = oauth.Consumer(key, secret)
        self.key = key
        self.secret = secret
        self.api_version = api_version
        self.signature = oauth.SignatureMethod_HMAC_SHA1()
        self.uri = "http://%s:%s" % (host, port)
        self.http = Http()
        self.headers = None

    def get_most_recent_http_headers(self):
        """ Intended for debugging -- return the most recent HTTP
        headers which were received from the server. """
        return self.headers

    def _endpoint(self, name, **kwargs):
        """Not used directly. Finds and formats the endpoints as needed for any type of request."""
        try:
            endpoint = self.endpoints[name]
        except KeyError:
            raise Exception('No endpoint named "%s"' % name)
        try:
            endpoint = endpoint % kwargs
        except KeyError, e:
            raise TypeError('Missing required argument "%s"' % (e.args[0],))
        return urljoin(urljoin(self.uri, self.api_version + '/'), endpoint)

    def get_feature(self, simplegeohandle):
        """Return the GeoJSON representation of a feature."""
        if not is_simplegeohandle(simplegeohandle):
            raise TypeError("simplegeohandle is required to match the regex %s, but it was %s :: %r" % (SIMPLEGEOHANDLE_RSTR, type(simplegeohandle), simplegeohandle))
        endpoint = self._endpoint('feature', simplegeohandle=simplegeohandle)
        return Feature.from_json(self._request(endpoint, 'GET')[1])

    def get_annotations(self, simplegeohandle):
        if not is_simplegeohandle(simplegeohandle):
            raise TypeError("simplegeohandle is required to match the regex %s, but it was %s :: %r" % (SIMPLEGEOHANDLE_RSTR, type(simplegeohandle), simplegeohandle))
        endpoint = self._endpoint('annotations', simplegeohandle=simplegeohandle)
        return json.loads(self._request(endpoint, 'GET')[1])

    def annotate(self, simplegeohandle, annotations, private):
        if not isinstance(annotations, dict):
            raise TypeError('annotations must be of type dict')
        if not len(annotations.keys()):
            raise ValueError('annotations dict is empty')
        for annotation_type in annotations.keys():
            if not len(annotations[annotation_type].keys()):
                raise ValueError('annotation type "%s" is empty' % annotation_type)
        if not isinstance(private, bool):
            raise TypeError('private must be of type bool')

        data = {'annotations': annotations,
                'private': private}

        endpoint = self._endpoint('annotations', simplegeohandle=simplegeohandle)
        return json.loads(self._request(endpoint,
                                        'POST',
                                        data=json.dumps(data))[1])

    def _request(self, endpoint, method, data=None):
        """
        Not used directly by code external to this lib. Performs the
        actual request against the API, including passing the
        credentials with oauth.  Returns a tuple of (headers as dict,
        body as string).
        """
        if data is not None:
            data = to_unicode(data)
        params = {}
        body = data
        request = oauth.Request.from_consumer_and_token(self.consumer,
            http_method=method, http_url=endpoint, parameters=params)

        request.sign_request(self.signature, self.consumer, None)
        headers = request.to_header(self.realm)
        headers['User-Agent'] = 'SimpleGeo Places Client v%s' % __version__

        self.headers, content = self.http.request(endpoint, method, body=body, headers=headers)

        if self.headers['status'][0] not in ('2', '3'):
            raise APIError(int(self.headers['status']), content, self.headers)

        return self.headers, content


class APIError(Exception):
    """Base exception for all API errors."""

    def __init__(self, code, msg, headers, description=''):
        self.code = code
        self.msg = msg
        self.headers = headers
        self.description = description

    def __str__(self):
        return self.__repr__()

    def __repr__(self):
        return "%s (#%s) %s" % (self.msg, self.code, self.description)

class DecodeError(APIError):
    """There was a problem decoding the API's response, which was
    supposed to be encoded in JSON, but which apparently wasn't."""

    def __init__(self, body, le):
        super(DecodeError, self).__init__(None, "Could not decode JSON from server.", None, repr(le))
        self.body = body

    def __repr__(self):
        return "%s content: %s" % (self.description, self.body)

def is_valid_ip(ip):
    try:
        ipaddr.IPAddress(ip)
    except ValueError:
        return False
    else:
        return True
